// @signumjs/core/out/api/factories/contract/publishContractByReference.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.publishContractByReference = void 0;
const constants_1 = require("../../../constants");
const signIfPrivateKey_1 = require("../../../internal/signIfPrivateKey");
const contracts_1 = require("@signumjs/contracts");
const contracts_2 = require("@signumjs/contracts");
/**
 * Use with {@link ApiComposer} and belongs to {@link ContractApi}.
 *
 * See details at {@link ContractApi.publishContractByReference}
*
* @category factories
*/
const publishContractByReference = (service) => (args) => (0, signIfPrivateKey_1.signIfPrivateKey)(service, args, async (a) => {
    const { dataHex } = (0, contracts_1.generateDataStack)(a.data || []);
    const feeNQT = a.feePlanck || (0, contracts_2.calculateMinimumCreationFee)({
        dataHex,
    }).getPlanck();
    const parameters = {
        deadline: a.deadline || constants_1.DefaultDeadline,
        description: a.description,
        feeNQT,
        referencedTransactionFullHash: a.referencedTransactionHash,
        name: a.name,
        publicKey: a.senderPublicKey,
        data: dataHex || undefined,
        broadcast: true,
    };
    return service.send('createATProgram', parameters);
});
exports.publishContractByReference = publishContractByReference;
