// @signumjs/core/out/api/factories/transaction/cancelSubscription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cancelSubscription = void 0;
const constants_1 = require("../../../constants");
const createParametersFromAttachment_1 = require("../../../internal/createParametersFromAttachment");
const signIfPrivateKey_1 = require("../../../internal/signIfPrivateKey");
/**
 * Use with {@link ApiComposer} and belongs to {@link TransactionApi}.
 *
 * See details at {@link TransactionApi.cancelSubscription}
*
* @category factories
*/
const cancelSubscription = (service) => (args) => (0, signIfPrivateKey_1.signIfPrivateKey)(service, args, async (a) => {
    let parameters = {
        subscription: a.subscriptionId,
        publicKey: a.senderPublicKey,
        feeNQT: a.feePlanck,
        deadline: a.deadline || constants_1.DefaultDeadline,
        referencedTransactionFullHash: a.referencedTransactionFullHash || undefined
    };
    if (a.attachment) {
        parameters = (0, createParametersFromAttachment_1.createParametersFromAttachment)(a.attachment, parameters);
    }
    return service.send('subscriptionCancel', parameters);
});
exports.cancelSubscription = cancelSubscription;
