// @signumjs/core/out/api/factories/transaction/sendAmountToSingleRecipient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendAmountToSingleRecipient = void 0;
const constants_1 = require("../../../constants");
const createParametersFromAttachment_1 = require("../../../internal/createParametersFromAttachment");
const signIfPrivateKey_1 = require("../../../internal/signIfPrivateKey");
const SmartContractPublickey = '0000000000000000000000000000000000000000000000000000000000000000';
/**
 * Use with {@link ApiComposer} and belongs to {@link TransactionApi}.
 *
 * See details at {@link TransactionApi.sendAmountToSingleRecipient}
*
* @category factories
*/
const sendAmountToSingleRecipient = (service) => (args) => (0, signIfPrivateKey_1.signIfPrivateKey)(service, args, async (a) => {
    let recipientPublicKey = a.recipientPublicKey || undefined;
    if (recipientPublicKey && recipientPublicKey === SmartContractPublickey) {
        recipientPublicKey = undefined;
    }
    let parameters = {
        amountNQT: a.amountPlanck,
        publicKey: a.senderPublicKey,
        recipient: a.recipientId,
        recipientPublicKey,
        feeNQT: a.feePlanck,
        deadline: a.deadline || constants_1.DefaultDeadline,
        referencedTransactionFullHash: a.referencedTransactionFullHash,
        skipAdditionalSecurityCheck: a.skipAdditionalSecurityCheck
    };
    if (a.attachment) {
        parameters = (0, createParametersFromAttachment_1.createParametersFromAttachment)(a.attachment, parameters);
    }
    return service.send('sendMoney', parameters);
});
exports.sendAmountToSingleRecipient = sendAmountToSingleRecipient;
