// @signumjs/core/out/constants/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionSmartContractSubtype = exports.TransactionAdvancedPaymentSubtype = exports.TransactionMiningSubtype = exports.TransactionArbitrarySubtype = exports.TransactionAssetSubtype = exports.TransactionLeasingSubtype = exports.TransactionMarketplaceSubtype = exports.TransactionPaymentSubtype = exports.TransactionType = exports.DefaultApiEndpoint = exports.DefaultDeadline = exports.AddressPrefix = exports.ApiVersion = void 0;
const apiVersion_1 = require("./apiVersion");
Object.defineProperty(exports, "ApiVersion", { enumerable: true, get: function () { return apiVersion_1.ApiVersion; } });
const transactionType_1 = require("./transactionType");
Object.defineProperty(exports, "TransactionType", { enumerable: true, get: function () { return transactionType_1.TransactionType; } });
const transactionArbitrarySubtype_1 = require("./transactionArbitrarySubtype");
Object.defineProperty(exports, "TransactionArbitrarySubtype", { enumerable: true, get: function () { return transactionArbitrarySubtype_1.TransactionArbitrarySubtype; } });
const transactionAssetSubtype_1 = require("./transactionAssetSubtype");
Object.defineProperty(exports, "TransactionAssetSubtype", { enumerable: true, get: function () { return transactionAssetSubtype_1.TransactionAssetSubtype; } });
const transactionLeasingSubtype_1 = require("./transactionLeasingSubtype");
Object.defineProperty(exports, "TransactionLeasingSubtype", { enumerable: true, get: function () { return transactionLeasingSubtype_1.TransactionLeasingSubtype; } });
const transactionMarketplaceSubtype_1 = require("./transactionMarketplaceSubtype");
Object.defineProperty(exports, "TransactionMarketplaceSubtype", { enumerable: true, get: function () { return transactionMarketplaceSubtype_1.TransactionMarketplaceSubtype; } });
const transactionPaymentSubtype_1 = require("./transactionPaymentSubtype");
Object.defineProperty(exports, "TransactionPaymentSubtype", { enumerable: true, get: function () { return transactionPaymentSubtype_1.TransactionPaymentSubtype; } });
const transactionMiningSubtype_1 = require("./transactionMiningSubtype");
Object.defineProperty(exports, "TransactionMiningSubtype", { enumerable: true, get: function () { return transactionMiningSubtype_1.TransactionMiningSubtype; } });
const transactionAdvancedPaymentSubtype_1 = require("./transactionAdvancedPaymentSubtype");
Object.defineProperty(exports, "TransactionAdvancedPaymentSubtype", { enumerable: true, get: function () { return transactionAdvancedPaymentSubtype_1.TransactionAdvancedPaymentSubtype; } });
const transactionSmartContractSubtype_1 = require("./transactionSmartContractSubtype");
Object.defineProperty(exports, "TransactionSmartContractSubtype", { enumerable: true, get: function () { return transactionSmartContractSubtype_1.TransactionSmartContractSubtype; } });
/**
 * The default deadline (in minutes) for Transactions
 *
 */
const DefaultDeadline = 1440;
exports.DefaultDeadline = DefaultDeadline;
/**
 * The default endpoint for {@link ApiSettings}
 *
 */
const DefaultApiEndpoint = '/api';
exports.DefaultApiEndpoint = DefaultApiEndpoint;
/**
 *
 * Address prefixes used in {@link Address}
 *
 * */
var AddressPrefix;
(function (AddressPrefix) {
    AddressPrefix["MainNet"] = "S";
    AddressPrefix["TestNet"] = "TS";
})(AddressPrefix || (exports.AddressPrefix = AddressPrefix = {}));
