// @signumjs/core/out/constants/transactionAdvancedPaymentSubtype.js
"use strict";
/**
 * Original work Copyright (c) 2019 Burst Apps Team
 * Modified 2024, Signum Network
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionAdvancedPaymentSubtype = void 0;
/**
 * Constants for advanced payment subtypes
 *
 * @category transaction-types
 */
var TransactionAdvancedPaymentSubtype;
(function (TransactionAdvancedPaymentSubtype) {
    TransactionAdvancedPaymentSubtype[TransactionAdvancedPaymentSubtype["EscrowCreation"] = 0] = "EscrowCreation";
    TransactionAdvancedPaymentSubtype[TransactionAdvancedPaymentSubtype["EscrowSigning"] = 1] = "EscrowSigning";
    TransactionAdvancedPaymentSubtype[TransactionAdvancedPaymentSubtype["EscrowResult"] = 2] = "EscrowResult";
    TransactionAdvancedPaymentSubtype[TransactionAdvancedPaymentSubtype["SubscriptionSubscribe"] = 3] = "SubscriptionSubscribe";
    TransactionAdvancedPaymentSubtype[TransactionAdvancedPaymentSubtype["SubscriptionCancel"] = 4] = "SubscriptionCancel";
    TransactionAdvancedPaymentSubtype[TransactionAdvancedPaymentSubtype["SubscriptionPayment"] = 5] = "SubscriptionPayment";
})(TransactionAdvancedPaymentSubtype || (exports.TransactionAdvancedPaymentSubtype = TransactionAdvancedPaymentSubtype = {}));
