// @signumjs/core/out/constants/transactionAssetSubtype.js
"use strict";
/**
 * Original work Copyright (c) 2019 Burst Apps Team
 * Modified work Copyright (c) 2023 Signum Network
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionAssetSubtype = void 0;
/**
 * Constants for asset/token subtypes
 *
 * @category transaction-types
 */
var TransactionAssetSubtype;
(function (TransactionAssetSubtype) {
    TransactionAssetSubtype[TransactionAssetSubtype["AssetIssuance"] = 0] = "AssetIssuance";
    TransactionAssetSubtype[TransactionAssetSubtype["AssetTransfer"] = 1] = "AssetTransfer";
    TransactionAssetSubtype[TransactionAssetSubtype["AskOrderPlacement"] = 2] = "AskOrderPlacement";
    TransactionAssetSubtype[TransactionAssetSubtype["BidOrderPlacement"] = 3] = "BidOrderPlacement";
    TransactionAssetSubtype[TransactionAssetSubtype["AskOrderCancellation"] = 4] = "AskOrderCancellation";
    TransactionAssetSubtype[TransactionAssetSubtype["BidOrderCancellation"] = 5] = "BidOrderCancellation";
    TransactionAssetSubtype[TransactionAssetSubtype["AssetMint"] = 6] = "AssetMint";
    TransactionAssetSubtype[TransactionAssetSubtype["AssetAddTreasureyAccount"] = 7] = "AssetAddTreasureyAccount";
    TransactionAssetSubtype[TransactionAssetSubtype["AssetDistributeToHolders"] = 8] = "AssetDistributeToHolders";
    TransactionAssetSubtype[TransactionAssetSubtype["AssetMultiTransfer"] = 9] = "AssetMultiTransfer";
    TransactionAssetSubtype[TransactionAssetSubtype["AssetTransferOwnership"] = 10] = "AssetTransferOwnership";
})(TransactionAssetSubtype || (exports.TransactionAssetSubtype = TransactionAssetSubtype = {}));
