// @signumjs/core/out/constants/transactionType.js
"use strict";
/**
 * Original work Copyright (c) 2019 Burst Apps Team
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionType = void 0;
/**
 * Constants for transaction types
 *
 * The transaction type is part of every {@link Transaction} object
 * and used to distinguish block data. Additionally, to the transaction type
 * a subtype is sent, that specifies the kind of transaction more detailly.
 *
 * @category transaction-types
 */
var TransactionType;
(function (TransactionType) {
    /**
     * {@link TransactionPaymentSubtype}
     */
    TransactionType[TransactionType["Payment"] = 0] = "Payment";
    /**
     * {@link TransactionArbitrarySubtype}
     */
    TransactionType[TransactionType["Arbitrary"] = 1] = "Arbitrary";
    /**
     * {@link TransactionAssetSubtype}
     */
    TransactionType[TransactionType["Asset"] = 2] = "Asset";
    /**
     * {@link TransactionMarketplaceSubtype}
     */
    TransactionType[TransactionType["Marketplace"] = 3] = "Marketplace";
    /**
     * {@link TransactionLeasingSubtype}
     */
    TransactionType[TransactionType["Leasing"] = 4] = "Leasing";
    /**
     * {@link TransactionMiningSubtype}
     */
    TransactionType[TransactionType["Mining"] = 20] = "Mining";
    /**
     * {@link TransactionAdvancedPaymentSubtype}
     */
    TransactionType[TransactionType["AdvancedPayment"] = 21] = "AdvancedPayment";
    /**
     * {@link TransactionSmartContractSubtype}
     */
    TransactionType[TransactionType["SmartContract"] = 22] = "SmartContract";
})(TransactionType || (exports.TransactionType = TransactionType = {}));
