// @signumjs/core/out/constants/transactionArbitrarySubtype.js
"use strict";
/**
 * Original work Copyright (c) 2019 Burst Apps Team
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionArbitrarySubtype = void 0;
/**
 * Constants for arbitrary subtypes
 *
 * @category transaction-types
 */
var TransactionArbitrarySubtype;
(function (TransactionArbitrarySubtype) {
    TransactionArbitrarySubtype[TransactionArbitrarySubtype["Message"] = 0] = "Message";
    TransactionArbitrarySubtype[TransactionArbitrarySubtype["AliasAssignment"] = 1] = "AliasAssignment";
    TransactionArbitrarySubtype[TransactionArbitrarySubtype["PollCreation"] = 2] = "PollCreation";
    TransactionArbitrarySubtype[TransactionArbitrarySubtype["VoteCasting"] = 3] = "VoteCasting";
    TransactionArbitrarySubtype[TransactionArbitrarySubtype["HubAnnouncement"] = 4] = "HubAnnouncement";
    TransactionArbitrarySubtype[TransactionArbitrarySubtype["AccountInfo"] = 5] = "AccountInfo";
    TransactionArbitrarySubtype[TransactionArbitrarySubtype["AliasSale"] = 6] = "AliasSale";
    TransactionArbitrarySubtype[TransactionArbitrarySubtype["AliasBuy"] = 7] = "AliasBuy";
    TransactionArbitrarySubtype[TransactionArbitrarySubtype["TopLevelDomainAssignment"] = 8] = "TopLevelDomainAssignment";
})(TransactionArbitrarySubtype || (exports.TransactionArbitrarySubtype = TransactionArbitrarySubtype = {}));
