// @signumjs/core/out/constants/transactionMarketplaceSubtype.js
"use strict";
/**
 * Original work Copyright (c) 2019 Burst Apps Team
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionMarketplaceSubtype = void 0;
/**
 * Constants for marketplace subtypes
 *
 * @category transaction-types
 */
var TransactionMarketplaceSubtype;
(function (TransactionMarketplaceSubtype) {
    TransactionMarketplaceSubtype[TransactionMarketplaceSubtype["MarketplaceListing"] = 0] = "MarketplaceListing";
    TransactionMarketplaceSubtype[TransactionMarketplaceSubtype["MarketplaceRemoval"] = 1] = "MarketplaceRemoval";
    TransactionMarketplaceSubtype[TransactionMarketplaceSubtype["MarketplaceItemPriceChange"] = 2] = "MarketplaceItemPriceChange";
    TransactionMarketplaceSubtype[TransactionMarketplaceSubtype["MarketplaceItemQuantityChange"] = 3] = "MarketplaceItemQuantityChange";
    TransactionMarketplaceSubtype[TransactionMarketplaceSubtype["MarketplacePurchase"] = 4] = "MarketplacePurchase";
    TransactionMarketplaceSubtype[TransactionMarketplaceSubtype["MarketplaceDelivery"] = 5] = "MarketplaceDelivery";
    TransactionMarketplaceSubtype[TransactionMarketplaceSubtype["MarketplaceFeedback"] = 6] = "MarketplaceFeedback";
    TransactionMarketplaceSubtype[TransactionMarketplaceSubtype["MarketplaceRefund"] = 7] = "MarketplaceRefund";
})(TransactionMarketplaceSubtype || (exports.TransactionMarketplaceSubtype = TransactionMarketplaceSubtype = {}));
