// @signumjs/core/out/api/factories/alias/buyTopLevelDomain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buyTopLevelDomain = void 0;
const signIfPrivateKey_1 = require("../../../internal/signIfPrivateKey");
const constants_1 = require("../../../constants");
const internal_1 = require("../../../internal");
/**
 * Use with {@link ApiComposer} and belongs to {@link AliasApi}.
 *
 * See details at {@link AliasApi.buyTopLevelDomain}
 *
*
* @category factories
*/
const buyTopLevelDomain = (service) => (args) => (0, signIfPrivateKey_1.signIfPrivateKey)(service, args, async (a) => {
    let parameters = {
        tld: a.tld,
        amountNQT: a.amountPlanck,
        deadline: a.deadline || constants_1.DefaultDeadline,
        feeNQT: a.feePlanck,
        publicKey: a.senderPublicKey,
        referencedTransactionFullHash: a.referencedTransactionFullHash,
    };
    if (args.attachment) {
        parameters = (0, internal_1.createParametersFromAttachment)(a.attachment, parameters);
    }
    return service.send('setTLD', parameters);
});
exports.buyTopLevelDomain = buyTopLevelDomain;
