// @signumjs/core/out/api/factories/asset/addAssetTreasuryAccount.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addAssetTreasuryAccount = void 0;
const constants_1 = require("../../../constants");
const createParametersFromAttachment_1 = require("../../../internal/createParametersFromAttachment");
const signIfPrivateKey_1 = require("../../../internal/signIfPrivateKey");
/**
 *
 * Use with {@link ApiComposer} and belongs to {@link AssetApi}.
 *
 * See details at {@link AssetApi.addAssetTreasuryAccount}
 *
*
* @category factories
*/
const addAssetTreasuryAccount = (service) => (args) => (0, signIfPrivateKey_1.signIfPrivateKey)(service, args, async (a) => {
    let parameters = {
        recipient: a.accountId,
        publicKey: a.senderPublicKey,
        recipientPublicKey: a.recipientPublicKey || undefined,
        referencedTransactionFullHash: a.referencedTransactionFullHash,
        feeNQT: a.feePlanck,
        deadline: a.deadline || constants_1.DefaultDeadline
    };
    if (a.attachment) {
        parameters = (0, createParametersFromAttachment_1.createParametersFromAttachment)(a.attachment, parameters);
    }
    return service.send('addAssetTreasuryAccount', parameters);
});
exports.addAssetTreasuryAccount = addAssetTreasuryAccount;
