// @signumjs/core/out/api/factories/asset/cancelOrder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cancelOrder = void 0;
const constants_1 = require("../../../constants");
const createParametersFromAttachment_1 = require("../../../internal/createParametersFromAttachment");
const signIfPrivateKey_1 = require("../../../internal/signIfPrivateKey");
/**
 * @ignore
 * This is an internal helper
 * See details at {@link AssetApi.cancelAskOrder} {@link AssetApi.cancelBidOrder}
*
* @category factories
*/
const cancelOrder = (service) => (args) => (0, signIfPrivateKey_1.signIfPrivateKey)(service, args, async (a) => {
    let parameters = {
        order: a.order,
        publicKey: a.senderPublicKey,
        feeNQT: a.feePlanck,
        deadline: a.deadline || constants_1.DefaultDeadline,
        referencedTransactionFullHash: a.referencedTransactionFullHash || undefined
    };
    if (a.attachment) {
        parameters = (0, createParametersFromAttachment_1.createParametersFromAttachment)(a.attachment, parameters);
    }
    const method = a.type === 'ask' ? 'cancelAskOrder' : 'cancelBidOrder';
    return service.send(method, parameters);
});
exports.cancelOrder = cancelOrder;
