// @signumjs/core/out/api/factories/asset/issueAsset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.issueAsset = void 0;
const constants_1 = require("../../../constants");
const createParametersFromAttachment_1 = require("../../../internal/createParametersFromAttachment");
const signIfPrivateKey_1 = require("../../../internal/signIfPrivateKey");
/**
 *
 * Use with {@link ApiComposer} and belongs to {@link AssetApi}.
 *
 * See details at {@link AssetApi.issueAsset}
 *
 * @category factories
 */
const issueAsset = (service) => (args) => (0, signIfPrivateKey_1.signIfPrivateKey)(service, args, async (a) => {
    let parameters = {
        name: a.name,
        description: a.description,
        quantityQNT: a.quantity,
        decimals: a.decimals,
        publicKey: a.senderPublicKey,
        feeNQT: a.feePlanck,
        mintable: a.mintable,
        deadline: a.deadline || constants_1.DefaultDeadline,
    };
    if (a.attachment) {
        parameters = (0, createParametersFromAttachment_1.createParametersFromAttachment)(a.attachment, parameters);
    }
    return service.send('issueAsset', parameters);
});
exports.issueAsset = issueAsset;
