// @signumjs/core/out/api/factories/asset/placeOrder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.placeOrder = void 0;
const constants_1 = require("../../../constants");
const createParametersFromAttachment_1 = require("../../../internal/createParametersFromAttachment");
const signIfPrivateKey_1 = require("../../../internal/signIfPrivateKey");
const util_1 = require("@signumjs/util");
/**
 * @ignore
 * This is an internal helper
 * See details at {@link AssetApi.placeAskOrder} {@link AssetApi.placeBidOrder}
*
* @category factories
*/
const placeOrder = (service) => (args) => (0, signIfPrivateKey_1.signIfPrivateKey)(service, args, async (a) => {
    let decimals = args.decimals;
    if (decimals === undefined) {
        const asset = await service.query('getAsset', { asset: args.assetId });
        decimals = asset.decimals;
    }
    let parameters = {
        asset: a.assetId,
        priceNQT: (0, util_1.convertPlanckToAssetPrice)(a.pricePlanck, decimals),
        quantityQNT: a.quantity,
        publicKey: a.senderPublicKey,
        feeNQT: a.feePlanck,
        deadline: a.deadline || constants_1.DefaultDeadline,
        referencedTransactionFullHash: a.referencedTransactionFullHash || undefined
    };
    if (a.attachment) {
        parameters = (0, createParametersFromAttachment_1.createParametersFromAttachment)(a.attachment, parameters);
    }
    const method = a.type === 'ask' ? 'placeAskOrder' : 'placeBidOrder';
    return service.send(method, parameters);
});
exports.placeOrder = placeOrder;
