// @signumjs/core/out/api/factories/asset/transferAsset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transferAsset = void 0;
const constants_1 = require("../../../constants");
const createParametersFromAttachment_1 = require("../../../internal/createParametersFromAttachment");
const signIfPrivateKey_1 = require("../../../internal/signIfPrivateKey");
/**
 *
 * Use with {@link ApiComposer} and belongs to {@link AssetApi}.
 *
 * See details at {@link AssetApi.transferAsset}
 *
*
* @category factories
*/
const transferAsset = (service) => (args) => (0, signIfPrivateKey_1.signIfPrivateKey)(service, args, async (a) => {
    let parameters = {
        asset: a.assetId,
        quantityQNT: a.quantity,
        publicKey: a.senderPublicKey,
        recipient: a.recipientId,
        recipientPublicKey: a.recipientPublicKey || undefined,
        feeNQT: a.feePlanck,
        amountNQT: a.amountPlanck || undefined,
        deadline: a.deadline || constants_1.DefaultDeadline,
        referencedTransactionFullHash: a.referencedTransactionFullHash,
        skipAdditionalSecurityCheck: a.skipAdditionalSecurityCheck
    };
    if (a.attachment) {
        parameters = (0, createParametersFromAttachment_1.createParametersFromAttachment)(a.attachment, parameters);
    }
    return service.send('transferAsset', parameters);
});
exports.transferAsset = transferAsset;
