// @signumjs/core/out/api/factories/asset/transferAssetOwnership.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transferAssetOwnership = void 0;
const constants_1 = require("../../../constants");
const createParametersFromAttachment_1 = require("../../../internal/createParametersFromAttachment");
const signIfPrivateKey_1 = require("../../../internal/signIfPrivateKey");
/**
 *
 * Use with {@link ApiComposer} and belongs to {@link AssetApi}.
 *
 * See details at {@link AssetApi.transferAssetOwnership}
 *
*
* @category factories
*/
const transferAssetOwnership = (service) => (args) => (0, signIfPrivateKey_1.signIfPrivateKey)(service, args, async (a) => {
    let parameters = {
        publicKey: a.senderPublicKey,
        recipient: a.recipientId,
        recipientPublicKey: a.recipientPublicKey || undefined,
        feeNQT: a.feePlanck,
        deadline: a.deadline || constants_1.DefaultDeadline,
        referencedTransactionFullHash: a.referencedTransactionFullHash
    };
    if (a.attachment) {
        parameters = (0, createParametersFromAttachment_1.createParametersFromAttachment)(a.attachment, parameters);
    }
    return service.send('transferAssetOwnership', parameters);
});
exports.transferAssetOwnership = transferAssetOwnership;
