// @signumjs/core/out/internal/createParametersFromAttachment.js
"use strict";
/** @ignore */
/** @internal */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createParametersFromAttachment = void 0;
const attachment_1 = require("../typings/attachment");
/**
 * Creates send parameters for a transaction from attachment data
 *
 * @internal
 * @param attachment The attachment
 * @param params Any object
 * @return HttpParams
 */
const createParametersFromAttachment = (attachment, params) => {
    if (attachment instanceof attachment_1.AttachmentEncryptedMessage) {
        return {
            encryptedMessageData: attachment.data,
            encryptedMessageNonce: attachment.nonce,
            messageToEncryptIsText: String(attachment.isText),
            ...params
        };
    }
    if (attachment instanceof attachment_1.AttachmentMessage) {
        return {
            message: attachment.message,
            messageIsText: String(attachment.messageIsText),
            ...params
        };
    }
    throw new Error(`Unknown attachment type: ${JSON.stringify(attachment)}`);
};
exports.createParametersFromAttachment = createParametersFromAttachment;
