// @signumjs/core/out/api/factories/message/sendMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendMessage = void 0;
const constants_1 = require("../../../constants");
const signIfPrivateKey_1 = require("../../../internal/signIfPrivateKey");
/**
 * Use with {@link ApiComposer} and belongs to {@link MessageApi}.
 *
 * See details at {@link MessageApi.sendMessage}
*
* @category factories
*/
const sendMessage = (service) => (args) => (0, signIfPrivateKey_1.signIfPrivateKey)(service, args, async (a) => {
    const parameters = {
        message: a.message,
        publicKey: a.senderPublicKey,
        recipient: a.recipientId,
        recipientPublicKey: a.recipientPublicKey || undefined,
        feeNQT: a.feePlanck,
        deadline: a.deadline || constants_1.DefaultDeadline,
        messageIsText: a.messageIsText !== false,
        broadcast: true,
        referencedTransactionFullHash: a.referencedTransactionFullHash || undefined
    };
    return service.send('sendMessage', parameters);
});
exports.sendMessage = sendMessage;
